File: 1170-compare-strings-by-frequency-of-the-smallest-character/1170-compare-strings-by-frequency-of-the-smallest-character.js
/**
 * @param {string[]} queries
 * @param {string[]} words
 * @return {number[]}
 */
var numSmallerByFrequency = function(queries, words) {
    
    let result = [];
    let freqCharWords = words.map((word) =>
    {
        return findFreq(word);
    });

    freqCharWords.sort((a, b) => a - b);

    for(let i = 0 ; i < queries.length ; i++)
    {
        let freq = findFreq(queries[i]);
        result.push(binarySearch(freqCharWords, freq));
    }

    return result;
};

function findFreq(word)
{
    let freq =  new Array(26).fill(0);

     word.split("").forEach(el =>
    {
        freq[el.charCodeAt(0) - 'a'.charCodeAt(0)]++
    })

    for(let num of freq)
    {
        if(num != 0)
            return num;
    }

    return -1;
}

function binarySearch(nums, target)
{
    let length = nums.length;
    let low = 0;
    let high = length - 1;

    while(low <= high)
    {
        let mid = Math.floor((low + high) / 2);

        if(nums[mid] <= target)
            low = mid + 1;
        else
            high = mid - 1;
    }

    return length - low;
}

File: 1170-compare-strings-by-frequency-of-the-smallest-character/1170-compare-strings-by-frequency-of-the-smallest-character.java
class Solution {
    public int[] numSmallerByFrequency(String[] queries, String[] words) {

        int n = queries.length;
        int m = words.length;
        int result[] = new int[n];
        int freqWords[] = new int[m];

        for(int i = 0 ; i < m ; i++)
        {
            freqWords[i] = findFreq(words[i]);
        }

        Arrays.sort(freqWords);

        for(int i = 0 ; i < n ; i++)
        {
            int freq = findFreq(queries[i]);
            result[i] = binarySearch(freqWords, freq);
        }

        return result;
    }

    public int findFreq(String s)
    {
        int freq[] = new int[26];

        for(char ch : s.toCharArray())
            freq[ch - 'a']++;

        for(int num : freq)
        {
            if(num != 0)
                return num;
        }

        return 0;
    }

    public int binarySearch(int nums[], int target)
    {
        int length = nums.length;
        int low = 0;
        int high = length - 1;

        while(low <= high)
        {
            int mid = (low + high) / 2;

            if(nums[mid] <= target)
                low = mid + 1;
            else
                high = mid - 1;
        }

        return length - low;
    }
}